File: csc454hw03/src/main/java/MemoryModel.java

/**
 * In the memory model, the state consists of a pair (x1, x2). The output of the model is x1, and the state transitions
 * on input x3, will result in (x2,x3).
 */
public class MemoryModel extends Model {

    public MemoryModel(int [] initial, Port in, Port out) {
        super(initial);
        this.outPort = out;
        this.numberOfInputs = 1;
        this.inPorts = new Port [numberOfInputs];
        this.inPorts[0] = in;
    }

    @Override
    public int lambda() {
        return this.state[0];
    }

    @Override
    public void delta(int [] inputSet) {
        int x1 = this.state[1];
        int x2 = inputSet[0];
        int [] stateVals = {x1, x2};
        this.state = stateVals;
    }

}

File: csc454hw04_cpp/MemoryModel.cpp
#ifndef MEMORY_MODEL
#define MEMORY_MODEL

#include "Model.cpp"
#include "Port.cpp"

class MemoryModel : public Model
{

public:
    MemoryModel(int *initialState, Port *in, Port *out)
    {
        state = new int [2];
        state[0] = initialState[0];
        state[1]= initialState[1];
        outPort = out;
        numberOfInputs = 1;
        inPorts = new Port *[numberOfInputs];
        inPorts[0] = in;
        stateSize = 2;
    }

    int lambda()
    {
        return state[0];
    }

    void delta(int *inputSet)
    {
        state[0] = state[1];
        state[1] = inputSet[0];
    }

    ~MemoryModel()
    {
        for (int i = 0; i < numberOfInputs; i++)
        {
            delete inPorts[i];
        }
        delete[] inPorts;
        delete state;
    }
};

#endif